bundles/org.openhab.binding.icloud/src/main/java/org/openhab/binding/icloud/internal/handler/ICloudDeviceHandler.java

```diff
@@ -91,35 +96,33 @@ public void deviceInformationUpdate(List<ICloudDeviceInformation> deviceInformat
 
     @Override
     public void initialize() {
-        Bridge bridge = getBridge();
-        Object bridgeStatus = (bridge == null) ? null : bridge.getStatus();
-        logger.debug("initializeThing thing [{}]; bridge status: [{}]", getThing().getUID(), bridgeStatus);
-
         ICloudDeviceThingConfiguration configuration = getConfigAs(ICloudDeviceThingConfiguration.class);
         this.deviceId = configuration.deviceId;
 
-        ICloudAccountBridgeHandler handler = getIcloudAccount();
-        if (handler != null) {
-            refreshData();
-        } else {
-            updateStatus(OFFLINE, BRIDGE_UNINITIALIZED, "Bridge not found");
-        }
-    }
+        Bridge bridge = getBridge();
+        Object bridgeStatus = (bridge == null) ? null : bridge.getStatus();
+        this.logger.debug("initializeThing thing [{}]; bridge status: [{}]", getThing().getUID(), bridgeStatus);
 
-    private void refreshData() {
-        ICloudAccountBridgeHandler bridge = getIcloudAccount();
         if (bridge != null) {
-            bridge.refreshData();
+            ((ICloudAccountBridgeHandler) bridge.getHandler()).registerListener(this);
+        }
+
+        if (bridge.getStatus() == ThingStatus.ONLINE) {
+            ((ICloudAccountBridgeHandler) bridge.getHandler()).refreshData();
+            updateStatus(ThingStatus.ONLINE);
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_UNINITIALIZED);
```

Comment:
Something here is still not correct. If bridge is null, you can exit immediately:

```java
Bridge bridge = getBridge();
if (bridge == null) {
    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Bridge is not configured");
    return;
}
```

After this point, you can set status to **UNKNOWN** if you cannot quickly determine if online or not.

If bridge is offline, **BRIDGE_OFFLINE** is the correct detail.

Reply:
> Something here is still not correct. If bridge is null, you can exit immediately:
> 
> ```java
> Bridge bridge = getBridge();
> if (bridge == null) {
>     updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Bridge is not configured");
>     return;
> }
> ```
> 
> After this point, you can set status to **UNKNOWN** if you cannot quickly determine if online or not.
> 
> If bridge is offline, **BRIDGE_OFFLINE** is the correct detail.

Think this improved with the commit i just did. I am calling it a day for now. I might proceed tomorrow. Depends on @maihacke what he wants, as i commit to his repository/PR. 

There are several other warnings to be fixed and i just drop in, so not sure

Reply:
Btw, see also #12330.

Reply:
This is still not resolved, so reopening. I'll try to add some more explanation.